### PR TITLE
Add HIP heFFTe, update FFT data types, and test with rocm 4.0

### LINKED ISSUE #319: Add interface for heFFTe HIP support
Extend FFT to use rocFFT backend with the heFFTe 2.0 release

### PR DESCRIPTION
**Pull request by streeve**
Closes #319 

This PR does two things. First, to simplify the FFT interface, Kokkos Views with a pair of float/double is cast to `std::complex` when passed to heFFTe (rather than separate device complex types and `Kokkos::complex`). This also enabled adding FFTs with HIP, using the heFFTe rocm backend.

Testing with `rocm 4.0` was added (the `-complete` docker image comes with `rocff`)

### CODE REVIEW

**Comment by streeve** [comment]
retest this please

**Comment by streeve** [comment]
retest this please

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 1]
```diff
@@ -301,7 +247,7 @@ class FastFourierTransform
 //---------------------------------------------------------------------------//
 // heFFTe
 //---------------------------------------------------------------------------//
-
+// TODO: dont think need Scalar as template param for these
```
Comment (dalg24) [inline cajita/src/Cajita_FastFourierTransform.hpp:250]: Are you planning to do it as part of this PR?

File: docker/Dockerfile.hipcc [hunk 2]
```diff
@@ -1,4 +1,4 @@
-ARG BASE=rocm/dev-ubuntu-18.04:3.5
+ARG BASE=rocm/dev-ubuntu-18.04:4.0-complete
```
[the remark this answers is not in the document]
Comment (streeve) [reply docker/Dockerfile.hipcc:1]: This is to get both `rocm` and `rocfft` together. I can add a comment to make that clear

**Comment by sfogerty** [comment]
retest this please

**Comment by streeve** [comment]
retest this please

File: docker/Dockerfile.hipcc [hunk 3]
```diff
@@ -24,6 +24,7 @@ RUN apt-get update && apt-get install -y \
     rm -rf /var/lib/apt/lists/*
 
 ENV PATH=/opt/rocm/bin:$PATH
+ENV ROCFFT_DIR=/opt/rocm/rocfft
```
[the remark this answers is not in the document]
Comment (streeve) [reply docker/Dockerfile.hipcc]: You're right - was not needed

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 1 above)
Comment (streeve) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Removed

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 4]
```diff
@@ -537,10 +407,9 @@ class HeffteFastFourierTransform
         // Create a subview of the work array to write the local data into.
         auto own_space =
             x.layout()->localGrid()->indexSpace( Own(), EntityType(), Local() );
-
-        auto work_view =
-            createView<complex_type, Kokkos::LayoutRight, DeviceType>(
-                own_space, _fft_work.data() );
+        auto work_view_space = appendDimension( own_space, 2 );
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:410]: nice

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 5]
```diff
@@ -553,17 +422,25 @@ class HeffteFastFourierTransform
                 auto iw = i - own_space.min( Dim::I );
                 auto jw = j - own_space.min( Dim::J );
                 auto kw = k - own_space.min( Dim::K );
-                work_view( iw, jw, kw ) = copyFromKokkosComplex(
-                    x_view( i, j, k, 0 ), work_view( iw, jw, kw ) );
+                work_view( iw, jw, kw, 0 ) = x_view( i, j, k, 0 );
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:425]: This is ultimately cleaner. I like this a lot.

File: cajita/src/Cajita_FastFourierTransform.hpp [hunk 6]
```diff
@@ -553,17 +422,25 @@ class HeffteFastFourierTransform
...
+            // _fft->forward( _fft_work.data(), _fft_work.data(), scale );
         }
         else if ( flag == -1 )
         {
-            _fft->backward( _fft_work.data(), _fft_work.data(), scale );
+            _fft->backward(
+                reinterpret_cast<std::complex<Scalar>*>( _fft_work.data() ),
+                reinterpret_cast<std::complex<Scalar>*>( _fft_work.data() ),
+                scale );
+            // _fft->backward( _fft_work.data(), _fft_work.data(), scale );
```
Comment (sslattery) [inline cajita/src/Cajita_FastFourierTransform.hpp:443]: Should this commented out code be removed?

File: docker/Dockerfile.hipcc [hunk 7]
```diff
@@ -104,3 +105,52 @@ RUN ARBORX_VERSION=4834bff44c23c9510c6ed93366638dcdf85ab217 && \
...
+    tar -xf ${HEFFTE_ARCHIVE} -C heffte --strip-components=1 && \
+    cd heffte && \
+    mkdir -p build && cd build && \
+    cmake \
+      -D CMAKE_INSTALL_PREFIX=${HEFFTE_DIR} \
+      -D CMAKE_PREFIX_PATH="${FFTW_DIR}" \
+      -D CMAKE_BUILD_TYPE=Debug \
+      -D CMAKE_CXX_COMPILER=hipcc \
+      -D CMAKE_CXX_STANDARD=14 \
+      -D CMAKE_CXX_FLAGS="--amdgpu-target=gfx906" \
```
Comment (sslattery) [inline docker/Dockerfile.hipcc:151]: Curious if Heffte has any built-in architure flags like Kokkos does

**Review by sslattery — APPROVED** [review]

File: cajita/src/Cajita_FastFourierTransform.hpp (hunk 6 above)
Comment (sfogerty) [reply cajita/src/Cajita_FastFourierTransform.hpp]: Removed. Thanks for catching that

File: docker/Dockerfile.hipcc (hunk 7 above)
Comment (streeve) [reply docker/Dockerfile.hipcc:151]: Miro mentioned that they plan to do something like that, but not yet

**Pull request merged**